npk/src/manifest.rs

```diff
@@ -206,49 +223,68 @@ enum MountSource {
         resource: String,
     },
     Tmpfs {
-        #[serde(deserialize_with = "deserialize_tmpfs_size")]
-        size: String,
+        #[serde(deserialize_with = "deserialize_tmpfs")]
+        tmpfs: u64,
     },
     Bind {
         host: PathBuf,
-        #[serde(default)]
-        flags: HashSet<MountFlag>,
-    },
-    Persist {
-        #[serde(default)]
+        #[serde(default, skip_serializing_if = "HashSet::is_empty")]
         flags: HashSet<MountFlag>,
     },
+    Dev(Dev),
+    Persist(Persist),
 }
 
-fn deserialize_tmpfs_size<'de, D>(deserializer: D) -> Result<String, D::Error>
+fn deserialize_tmpfs<'de, D>(deserializer: D) -> Result<u64, D::Error>
 where
     D: Deserializer<'de>,
 {
     struct SizeVisitor;
 
     impl<'de> Visitor<'de> for SizeVisitor {
-        type Value = String;
+        type Value = u64;
         fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
             formatter.write_str("a number of bytes or a string with the size (e.g. 25M)")
         }
 
         fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E> {
-            Ok(format!("{}", v))
+            Ok(v)
         }
 
         fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
         where
             E: serde::de::Error,
         {
-            // check if the string is correctly formatted as size in tmpfs(5)
             lazy_static! {
-                static ref TMPFS_SIZE: regex::Regex =
-                    regex::Regex::new(r"^\d+(?i:k|m|g)?$").expect("Wrong regex");
+                static ref RE: regex::Regex =
+                    regex::Regex::new(r"^(?P<value>\d+)(?P<unit>k|m|g)?$").expect("Invalid regex");
             }
-            if TMPFS_SIZE.is_match(v) {
-                Ok(v.to_owned())
+
+            let caps = RE
+                .captures(&v)
+                .ok_or_else(|| serde::de::Error::custom(format!("Invalid tmpfs size: {}", v)))?;
+
+            let value = caps
+                .name("value")
+                .unwrap()
+                .as_str()
+                .parse::<u64>()
+                .map_err(serde::de::Error::custom)?;
+            if let Some(unit) = caps.name("unit") {
+                let value = match unit.as_str() {
+                    "k" => value * 1024,
+                    "m" => value * 1024 * 1024,
+                    "g" => value * 1024 * 1024 * 1024,
+                    _ => {
+                        return Err(serde::de::Error::custom(format!(
+                            "Invalid tmpfs unit: {}",
+                            unit.as_str()
+                        )))
+                    }
+                };
+                Ok(value)
```

Comment:
I think this breaks the serialization/deserialization symmetry as we once discussed

Reply:
That's true. The problem is not solvable unless we want to have numbers in quotes (strings) in the manifest as you pointed out some weeks ago. The comparison of manifests in yaml needs to be done the numbers must be encoded in bytes. A option would be to remove the unit sugar...

Reply:
but why not simply let mount handle the format instead of adding the extra logic?